ksqldb-engine/src/main/java/io/confluent/ksql/internal/UtilizationMetricsListener.java

```diff
@@ -0,0 +1,147 @@
+package io.confluent.ksql.internal;
+
+import io.confluent.ksql.engine.QueryEventListener;
+import io.confluent.ksql.metastore.MetaStore;
+import io.confluent.ksql.services.ServiceContext;
+import io.confluent.ksql.util.QueryMetadata;
+import org.apache.kafka.common.utils.Time;
+import org.apache.kafka.streams.KafkaStreams;
+import org.apache.kafka.streams.processor.ThreadMetadata;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.BufferedReader;
+import java.io.FileReader;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+public class UtilizationMetricsListener implements Runnable, QueryEventListener {
+
+    private final String STREAM_THREAD_GROUP = "stream-thread-metrics";
+    private final String THREAD_ID = "thread-id";
+
+    private final List<KafkaStreams> kafkaStreams;
+    private final Logger logger = LoggerFactory.getLogger(UtilizationMetricsListener.class);
+    private final List<String> metrics;
+    private final Time time;
+    private final double windowSize;
+
+    private final Map<String, Double> previousPollTime;
+    private final Map<String, Double> previousRestoreConsumerPollTime;
+    private final Map<String, Double> previousSendTime;
+    private final Map<String, Double> previousFlushTime;
+
+    public UtilizationMetricsListener(final double windowSize){
+        this.kafkaStreams = new ArrayList<>();
+        this.metrics = new LinkedList<>();
+        // we can add these here or pass it in through the constructor
+        metrics.add("poll-time-total");
+        metrics.add("restore-consumer-poll-time-total");
+        metrics.add("send-time-total");
+        metrics.add("flush-time-total");
+        time = Time.SYSTEM;
+        this.windowSize = windowSize;
+        previousPollTime = new HashMap<>();
+        previousRestoreConsumerPollTime = new HashMap<>();
+        previousSendTime = new HashMap<>();
+        previousFlushTime = new HashMap<>();
+    }
+
+    @Override
+    public void onCreate(
+            final ServiceContext serviceContext,
+            final MetaStore metaStore,
+            final QueryMetadata queryMetadata) {
+        kafkaStreams.add(queryMetadata.getKafkaStreams());
+    }
+
+    @Override
+    public void onDeregister(final QueryMetadata query) {
+        // Question - if we terminate a query and then restart it, will the underling
+        // kafka streams have a new name? if so, do we want to remove everything from the
+        // previous value hashmaps? Or do we want to see that historical information still?
+        // Seems like if there's a chance we could re-use the name we'd want a clean slate
+        kafkaStreams.remove(query.getKafkaStreams());
+    }
+
+    @Override
+    public void run() {
+        logger.info("Reporting CSU thread level metrics");
```

Comment:
You'll probably get a more accurate measurement if you actually store the observed sample times rather than assuming the windowSize each time. So something like:

```
public void run() {
    while (true) {
        logger.info("the current processing ratio is " + processingRatio() + "%");
        Thread.sleep(windowSize);
    }
}

public double processingRatio() {
    long sampleTime = time.milliseconds();
    double blockedTime = sampleTime - lastSampleTime;
    long windowStart = lastSampleTime;
    long windowSize = sampleTime - lastSampleTime;
    ...
    lastSampleTime = sampleTime;
}
```

This way if the interval between runs is a little longer than you requested (which is more likely under heavy load) your computation won't be off.

Reply:
So we compute the processing ratio at the start of the window, and then sleep for the full window size, at which point the thread should call `run()` again, right? Won't those overlap in the ideal case (when everything is running on-time)?

I get storing sample time instead of assuming window size but I don't get the `Thread.sleep(windowSize);`

Reply:
sorry - I wrote this comment before I saw you were using a ScheduledExecutorService. You should just be able to do:
```
public void run() {
    logger.info("the current processing ratio is " + processingRatio() + "%");
}
```
My main point was about using the _observed_ difference between sample times rather than the requested difference (the interval we pass to the executor service).